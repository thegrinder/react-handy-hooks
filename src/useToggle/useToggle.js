import { useState, useMemo } from 'react';

const useToggle = (initial = false) => {
  const [on, set] = useState(initial);
  const actions = {
    toggle: () => set((prevState) => !prevState),
    toggleOn: () => set(true),
    toggleOff: () => set(false),
  };
  const value = useMemo(() => [on, actions], [actions, on]);
  return value;
};

export default useToggle;
