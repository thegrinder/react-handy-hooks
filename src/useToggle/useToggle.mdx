---
name: useToggle
menu: hooks
---

# useToggle

useToggle is a simple hook that returns a boolean value and three actions to update it.

## State

Boolean, `false` by default. You can provide the initial value by calling the hook with an argument, e.g `useToggle(true)`.

## Actions

### toggle

- toggles the current state value

### toggleOn

- sets the state value to `true`

### toggleOff

- sets the state value to `false`

## Example

import { Playground } from 'docz';
import { useToggle } from '../';

<Playground>
  {() => {
    const UseToggleDemo = () => {
      const [state, actions] = useToggle();
      return <button onClick={actions.toggle}>{state ? 'on' : 'off'}</button>;
    };
    return <UseToggleDemo />;
  }}
</Playground>
