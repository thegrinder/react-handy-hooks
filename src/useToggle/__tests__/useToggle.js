import { renderHook, act } from '@testing-library/react-hooks';
import useToggle from '../useToggle';

describe('useToggle', () => {
  it('should return the correct array', () => {
    const { result } = renderHook(() => useToggle());
    expect(result.current).toMatchSnapshot();
  });

  it('should initialize with a given argument', () => {
    const initial = true;
    const { result } = renderHook(() => useToggle(initial));
    expect(result.current[0]).toEqual(initial);
  });

  it('should initialize with false by default', () => {
    const { result } = renderHook(() => useToggle());
    expect(result.current[0]).toEqual(false);
  });

  describe('actions', () => {
    it('should handle toggling', () => {
      const { result } = renderHook(() => useToggle());
      expect(result.current[0]).toEqual(false);
      act(() => {
        result.current[1].toggle();
      });
      expect(result.current[0]).toEqual(true);
      act(() => {
        result.current[1].toggle();
      });
      expect(result.current[0]).toEqual(false);
    });

    it('should handle toggling on', () => {
      const { result } = renderHook(() => useToggle());
      expect(result.current[0]).toEqual(false);
      act(() => {
        result.current[1].toggleOn();
      });
      expect(result.current[0]).toEqual(true);
    });

    it('should handle toggling off', () => {
      const { result } = renderHook(() => useToggle(true));
      expect(result.current[0]).toEqual(true);
      act(() => {
        result.current[1].toggleOff();
      });
      expect(result.current[0]).toEqual(false);
    });
  });
});
