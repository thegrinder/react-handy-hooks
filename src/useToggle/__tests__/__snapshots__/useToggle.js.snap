// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useToggle should return the correct array 1`] = `
Array [
  false,
  Object {
    "toggle": [Function],
    "toggleOff": [Function],
    "toggleOn": [Function],
  },
]
`;
